### ISSUE TITLE
Unequal horizontal space between symbols in MathJax v4 + macOS + Chrome (or Safari)

### ISSUE THREAD

**Issue by AlexEdgcomb** [mathjax/MathJax#3005]
### Issue Summary

Unequal horizontal space between symbols on MathJax 4.0.0-alpha.1, macOS 12.6, and Chrome 110.0.5481.100 (or Safari 16.0).

Less noticeable unequal space in: Windows 11 + Chrome 110 and Windows 11 + Edge 110.

Spacing equal in: macOS 12.6 + Firefox 110 and Windows 11 + Firefox 110.

### Steps to Reproduce:

1. Bootup MacOS 12.6
2. Open Chrome 110
3. Go to https://codepen.io/alexedgcomb/pen/ZEMWVYM

Observed: Unequal spacing between `p` and `&and;`, than between `&and;` and `q`.
<img width="50" alt="image" src="https://user-images.githubusercontent.com/3512898/220690954-8b02a552-0fe4-4027-a2fb-abbf33ed2c87.png">
Expected: Equivalent spacing between symbols
<img width="45" alt="image" src="https://user-images.githubusercontent.com/3512898/220691121-16c7eed6-2530-4766-8730-980979ddd9cb.png">

### Technical details:

* MathJax Version: 4.0.0-alpha.1
* macOS 12.6
* Chrome 110.0.5481.100 (or Safari 16.0)

I am not using MathJax configuration and loading MathJax via
``` html
<script src="https://cdn.jsdelivr.net/npm/mathjax-full@4.0.0-alpha.1/es5/tex-mml-svg.min.js"></script>
```

### Supporting information:

After typesetting, DOM has an SVG with `p`, then an mjx-break, then a second SVG with `&and; q`.

Different OS and browser combos yield a different font-size for the mjx-break. Below, I've documented various combos using browserstack.com, sorted by mjx-break's font-size.

#### macOS 12.6 and Chrome 110
<img width="48" alt="image" src="https://user-images.githubusercontent.com/3512898/220697511-42b95955-d8f8-4c9e-b520-8d7cf3e1efa5.png">
mjx-break: `font-size: 9px`

#### macOS 12.6 and Safari 16
<img width="45" alt="image" src="https://user-images.githubusercontent.com/3512898/220700238-ae661df7-d487-4560-a44d-b9ec5b1d096a.png">
mjx-break: `font-size: 9px`

#### Windows 11 and Chrome 110
<img width="44" alt="image" src="https://user-images.githubusercontent.com/3512898/220692606-f837598a-395b-42b0-a3e9-b1e7e184c617.png">
mjx-break: `font-size: 6px`

#### Windows 11 and Edge 110
<img width="41" alt="image" src="https://user-images.githubusercontent.com/3512898/220695946-fa826231-4fcc-4243-81db-26b9fdfe9525.png">
mjx-break: `font-size: 6px`

#### Windows 11 and Firefox 110
<img width="42" alt="image" src="https://user-images.githubusercontent.com/3512898/220695206-13bdfc09-7d7a-4c01-bc99-0e317d728d03.png">
mjx-break: `font-size: 3.552px`

#### macOS 12.6 and Firefox 110
<img width="39" alt="image" src="https://user-images.githubusercontent.com/3512898/220692402-3470a5c5-1267-40cd-a97f-dedfef914ca5.png">
mjx-break: `font-size: 3.552px`

**Comment by dpvc** [comment]
Thanks for the report.  The `mjx-break` elements are used to mark potential breakpoints for inline-math line breaks.  They are supposed to be scaled to the proper size for the space at the breakpoint, but it looks like the scaling is running into the browser's minimum font size.

We can probably rework this to use a smaller space that gets enlarged instead of an em-sized space that gets reduced.  That should avoid the minimum font size issue.

For now, you may want to turn off the in-line breaking either in the MathJax contextual menu (uncheck Math Settings -> Wide Expressions -> Allow Inline Breaks), or change the MathJax configuration in the page:

``` javascript
MathJax = {
  output: {
    linebreaks: {
      inline: false
    }
  }
};
```

**Comment by AlexEdgcomb** [comment]
@dpvc , thank you! Would you happen to have a suggestion in MathJax 4 for both (1) avoiding the unequal horizontal space in Chrome and (2) left-aligning LaTeX that has a line break?

Repro step:
1. Bootup MacOS 12.6
2. Open Chrome
3. Go to https://codepen.io/alexedgcomb/pen/ZEMWVYM

Observed: (1), but the line are centered with each other and centered on the webpage.
<img width="886" alt="image" src="https://user-images.githubusercontent.com/3512898/221631044-bb0b2476-c9b4-4b86-8a0d-2b1d79b902be.png">

Expected: (1), and the lines are left-aligned with each other and left-aligned on the webpage, like in MJ 2.7.7:
https://codepen.io/alexedgcomb/pen/yLxVGmK
<img width="390" alt="image" src="https://user-images.githubusercontent.com/3512898/221627519-a0a5ff49-bfef-42ca-817c-32c7894e0a7f.png">

**Comment by dpvc** [comment]
OK, the problem is that the v4 line-break code treats the forced inline break as a display break and uses the `displayAlign` and `displayIndent` values to typeset the results when automatic inline breaks are off.  That's a bug, so thanks for pointing it out.

I will make a PR to fix it, but here is a configuration you can use for now until the next release:

``` javascript
  MathJax = {
    output: {
      linebreaks: {
        inline: false
      }
    },
    startup: {
      ready() {
        const {CommonWrapper} = MathJax._.output.common.Wrapper;
        Object.assign(CommonWrapper.prototype, {
          _processIndent: CommonWrapper.prototype.processIndent,
          processIndent(indentalign, indentshift, align = '', shift = '', width = this.metrics.containerWidth) {
            if (!this.jax.math.display) return ['left', 0];
            return this._processIndent(indentalign, indentshift, align, shift, width);
          }
        });
        MathJax.startup.defaultReady();
      }
    }
  };
```

I think that will get you the v2.7 results that you are looking for.

**Comment by dpvc** [comment]
I've been experimenting with the inline breaks, and have an alternative approach to handling the breakable spaces that should avoid the font-size issue of the current approach.  It uses `letter-spacing` to adjust the space size rather than `font-size`, and that seems to work better across browsers.

 Here is a configuration that implements this idea:

``` js
  MathJax = {
    output: {
      linebreaks: {
        inline: true
      }
    },
    startup: {
      ready() {
        const {SvgMath} = MathJax._.output.svg.Wrappers.math;
        Object.assign(SvgMath.styles, {
          'mjx-break::after': {
            content: '" "',
            'white-space': 'normal',
            'line-height': 0,
            'font-family': 'MJX-ZERO'
          },
          'mjx-break[size="1"]': {
            'letter-spacing': '-.889em'
          },
          'mjx-break[size="2"]': {
            'letter-spacing': '-.833em'
          },
          'mjx-break[size="3"]': {
            'letter-spacing': '-.778em'
          },
          'mjx-break[size="4"]': {
            'letter-spacing': '-.772em'
          },
          'mjx-break[size="5"]': {
            'letter-spacing': '-.667em'
          }
        });
        delete SvgMath.styles['mjx-break'];
        MathJax.startup.defaultReady();
      }
    }
  };
```

I'd appreciate it if you would try this out and see if that works for your situation, as it helps to have a real-world test.  Use this in place of the configuration from my previous comment.

This should handle all the usual spacing in TeX, though it doesn't handle explicit `lspace` and `rspace` attributes on MathML `mo` nodes (but those are rare in TeX input).

**Comment by AlexEdgcomb** [comment]
@dpvc , the letter-spacing approach has passed visual tests so far:

1. Using my previous test case (https://codepen.io/alexedgcomb/pen/JjaWLWG) against various browser combinations:
macOS 13.2 and Chrome 110
macOS 12.6 and Chrome 110
macOS 12.6 and Firefox 110
macOS 12.6 and Safari 16
Windows 11 and Chrome 110
Windows 11 and Firefox 110
Windows 11 and Edge 110

2. Sampled from various real-world math-heavy content, including energy balances, discrete math, and calculus-based circuits on macOS 13.2 and Chrome 110. I'll repeat the sampling on other combinations tomorrow.

**Comment by AlexEdgcomb** [comment]
@dpvc , completing item 2 testing: Letter-spacing approach passed visual sampling on the same content for all tested combinations:
macOS 13.2 and Firefox 110
macOS 13.2 and Safari 16
Windows 11 and Edge 110

A question popped into my head: Why is `\(p&and;q\)` typeset with 3 elements (svg, mtx-break, svg) whereas `\(p&and;q\colon\)` is typeset with 1 element (svg)?

**Comment by dpvc** [comment]
Thanks for the updated information.  This all looks good.

As for the two equations you mentioned, the in-line breaking isn't working entirely properly in 4.0.0-alpha.1.  In response to this issue tracker, I've been working on an update to the inline breaking to implement the `letter-spacing` approach, and found a number of problems with the marking of breakpoints as the algorithm currently stands.  I'm not sure which one causes it to miss the breakpoint in `p&and;q\colon` expression, but my current version does get the breakpoint right in that case, so it should be fixed in the next release, along with a number of other shortcomings.

Thanks for the reports and the extra testing.  That is much appreciated.

**Comment by AlexEdgcomb** [comment]
@dpvc , turns out, we have a case wherein we'd like to disable inline line breaks, which you've fixed here:
https://github.com/mathjax/MathJax-src/pull/916

Would you happen to have a suggestion for including that fix in 4.0.0-alpha.1 via the MathJax object?

**Comment by dpvc** [comment]
> Would you happen to have a suggestion for including that fix in 4.0.0-alpha.1 via the MathJax object?

Sure.  Add

``` js
const {CommonWrapper} = MathJax._.output.common.Wrapper;
Object.assign(CommonWrapper.prototype, {
  _processIndent: CommonWrapper.prototype.processIndent,
  processIndent(indentalign, indentshift, align = '', shift = '', width = this.metrics.containerWidth) {
    if (!this.jax.math.display) return ['left', 0];
    return this._processIndent(indentalign, indentshift, align, shift, width);
  }
});
```

to your `startup.ready` function.

Can I ask what the motivation is for not using the in-line breaking?  Note that you can prevent automatic inline breaks by putting braces around the expression, as in `${x+y+z}$`, but forced breaks will still occur.

**Comment by AlexEdgcomb** [comment]
@dpvc , thank you for the snippet and thanks for asking! Also, thank you for alerting me to the brace notation's impact on formatting.

I'm helping convert existing webpages built with MathJax 2 to use MathJax 4. Some ~350px containers have inline expressions toward the end of a line that move entirely to the next line, rather than wrap. Moving to the next line happens to be preferred.

For example:
`The discharge coefficient is calculated with \(C_d=C_\infty+\dfrac b{\mathrm{Re}^n}\). But, \(C_L=\dfrac{<!--\mathcal L-->\mathscr {L}}{{\displaystyle\frac12}\rho U^2A}\) is used to calculate a force, not to show a relationship between flow and head loss.`

The first expression wraps at the `=`, but we'd prefer the expression to entirely move down.

To the extent possible at aligning MJ2 and MJ4 formatting, that'd save time with adjusting each such case.

I see the value in being explicit in our LaTeX, though, by adding `{...}` instead. Thank you!
